src/lib.rs

```diff
@@ -113,7 +113,8 @@ impl<T> AsyncMutex<T> {
         G: Future<Item = (T, O), Error = (Option<T>, E)>,
         B: IntoFuture<Item = G::Item, Error = G::Error, Future = G>,
     {
-        match mem::replace(&mut self.inner.borrow_mut().resource, ResourceState::Empty) {
+        let resource = mem::replace(&mut self.inner.borrow_mut().resource, ResourceState::Empty);
+        match resource {
```

Comment:
Nice fix here!
I think that the reason this problem happens is that self.inner keeps being mutable borrowed for the whole match clause if we don't separate it.